Source/Core/Core/HW/GBACore.cpp

```diff
@@ -0,0 +1,517 @@
+    std::copy(buffer, buffer + 5, command.buffer.begin());
+
+  if (m_thread)
+  {
+    std::lock_guard<std::mutex> lock(m_queue_mutex);
+    m_command_queue.push(command);
+    m_idle = false;
+    m_command_cv.notify_one();
+  }
+  else
+  {
+    RunCommand(command);
+  }
+}
+
+std::vector<u8> Core::GetJoybusResponse()
+{
+  if (!m_running)
+    return {};
+
+  if (m_thread)
+  {
+    std::unique_lock<std::mutex> lock(m_response_mutex);
+    m_response_cv.wait(lock, [&] { return m_response_ready; });
+  }
+  m_response_ready = false;
+  return m_response;
+}
+
+void Core::Flush()
+{
+  if (!m_running || !m_thread)
+    return;
+  std::unique_lock<std::mutex> lock(m_queue_mutex);
+  m_response_cv.wait(lock, [&] { return m_idle; });
+}
+
+void Core::ThreadLoop()
+{
+  Common::SetCurrentThreadName(fmt::format("GBA{}", m_device_number + 1).c_str());
+  std::unique_lock<std::mutex> queue_lock(m_queue_mutex);
+  while (true)
+  {
+    m_command_cv.wait(queue_lock, [&] { return !m_command_queue.empty() || m_exit_loop; });
+    if (m_exit_loop)
+      break;
+    Command command{m_command_queue.front()};
+    m_command_queue.pop();
+    queue_lock.unlock();
+
+    RunCommand(command);
+
+    queue_lock.lock();
+    if (m_command_queue.empty())
+      m_idle = true;
+    m_response_cv.notify_one();
+  }
+}
+
+void Core::RunCommand(Command& command)
+{
+  RunUntil(command.ticks);
+  if (!command.sync_only)
+  {
+    u64 run_ahead = 0;
+    m_response.clear();
+    if (m_link_enabled)
+    {
+      int recvd = GBASIOJOYSendCommand(
+          &m_sio_driver, static_cast<GBASIOJOYCommand>(command.buffer[0]), &command.buffer[1]);
+      std::copy(command.buffer.begin() + 1, command.buffer.begin() + 1 + recvd,
+                std::back_inserter(m_response));
+      run_ahead = m_response.size() * SystemTimers::GetTicksPerSecond() / BYTES_PER_SECOND;
+    }
+
+    if (m_thread && !m_response_ready)
+    {
+      std::lock_guard<std::mutex> response_lock(m_response_mutex);
+      m_response_ready = true;
+      m_response_cv.notify_one();
+    }
+    else
+    {
+      m_response_ready = true;
+    }
+
+    if (run_ahead)
+      RunFor(run_ahead);
+  }
+}
+
+void Core::RunUntil(u64 gc_ticks)
+{
+  const u64 gc_frequency = SystemTimers::GetTicksPerSecond();
+  mTimingSchedule(
+      m_core->timing, &m_event,
+      static_cast<s32>((gc_ticks - m_last_gc_ticks) * GBA_ARM7TDMI_FREQUENCY / gc_frequency));
+  m_waiting_for_event = true;
+
+  s32 start_time = mTimingCurrentTime(m_core->timing);
+  while (m_waiting_for_event)
+    m_core->runLoop(m_core);
+  s32 end_time = mTimingCurrentTime(m_core->timing);
+
+  u64 d = (static_cast<u64>(end_time - start_time) * gc_frequency) + m_gc_ticks_remainder;
+  m_last_gc_ticks += d / GBA_ARM7TDMI_FREQUENCY;
+  m_gc_ticks_remainder = d % GBA_ARM7TDMI_FREQUENCY;
+}
+
+void Core::RunFor(u64 gc_ticks)
+{
+  RunUntil(m_last_gc_ticks + gc_ticks);
+}
+
+void Core::DoState(PointerWrap& p)
+{
+  Flush();
+
+  p.Do(m_video_buffer);
```

Comment:
Any chance to save states should also increment the version over in `State.cpp`:
https://github.com/dolphin-emu/dolphin/blob/bcd572a820427af472147f6238e907ec2ec2cab4/Source/Core/Core/State.cpp#L77